src/validation.cpp

```diff
@@ -1295,36 +1295,41 @@ bool MemPoolAccept::ConsensusScriptChecks(const ATMPArgs& args, Workspace& ws)
     return true;
 }
 
-bool MemPoolAccept::FinalizeSubpackage(const ATMPArgs& args, std::vector<Workspace>& workspaces)
+void MemPoolAccept::FinalizeSubpackage(const ATMPArgs& args)
 {
     AssertLockHeld(cs_main);
     AssertLockHeld(m_pool.cs);
-    const CTransaction& tx = *workspaces.front().m_ptx;
-    const uint256& hash = workspaces.front().m_hash;
-    TxValidationState& state = workspaces.front().m_state;
-    const bool bypass_limits = args.m_bypass_limits;
 
     if (!m_subpackage.m_all_conflicts.empty()) Assume(args.m_allow_replacement);
     // Remove conflicting transactions from the mempool
     for (CTxMemPool::txiter it : m_subpackage.m_all_conflicts)
     {
-        LogDebug(BCLog::MEMPOOL, "replacing mempool tx %s (wtxid=%s, fees=%s, vsize=%s). New tx %s (wtxid=%s, fees=%s, vsize=%s)\n",
+        LogDebug(BCLog::MEMPOOL, "replacing mempool tx %s (wtxid=%s, fees=%s, vsize=%s). ",
                 it->GetTx().GetHash().ToString(),
                 it->GetTx().GetWitnessHash().ToString(),
                 it->GetFee(),
-                it->GetTxSize(),
-                hash.ToString(),
-                tx.GetWitnessHash().ToString(),
-                workspaces[0].m_tx_handle->GetFee(),
-                workspaces[0].m_tx_handle->GetTxSize());
+                it->GetTxSize());
+        FeeFrac feerate = m_subpackage.m_changeset->GetAggregateFeeRate();
+        if (m_subpackage.m_changeset->GetTxCount() == 1) {
+            const CTransaction& tx = m_subpackage.m_changeset->GetTx(0);
+            LogDebug(BCLog::MEMPOOL, "New tx %s (wtxid=%s, fees=%s, vsize=%s)\n",
+                    tx.GetHash().ToString(),
+                    tx.GetWitnessHash().ToString(),
+                    feerate.fee, feerate.size);
+        } else {
+            LogDebug(BCLog::MEMPOOL, "New package with %lu txs, fees=%s, vsize=%s\n",
+                    m_subpackage.m_changeset->GetTxCount(),
+                    feerate.fee,
+                    feerate.size);
+        }
         TRACE7(mempool, replaced,
                 it->GetTx().GetHash().data(),
                 it->GetTxSize(),
                 it->GetFee(),
                 std::chrono::duration_cast<std::chrono::duration<std::uint64_t>>(it->GetTime()).count(),
-                hash.data(),
-                workspaces[0].m_tx_handle->GetTxSize(),
-                workspaces[0].m_tx_handle->GetFee()
+                m_subpackage.m_changeset->GetTx(0).GetHash().data(),
```

Comment:
Why are we only tracing the first tx in the package

Reply:
Not sure if the tracing API is well-specified for package based replacement.  I'm open to doing something different than what I propose here though, suggestions?

Reply:
Its not specified, I think we have to specify tracing api for packages;

Should be something like

#### Tracepoint `mempool:replaced_by_package`

Is called when a transaction in the node's mempool is getting replaced by a package.
Passes information about the replaced transaction and  the replacement package.

Arguments passed:
1. Replaced transaction ID (hash) as `pointer to unsigned chars` (i.e. 32 bytes in little-endian)
2. Replaced transaction virtual size as `int32`
3. Replaced transaction fee as `int64`
4. Replaced transaction mempool entry time (epoch) as `uint64`
5. Replacement package virtual size as `int32`
6. Replacement package fee as `int64`

Reply:
Tracking this issue above (https://github.com/bitcoin/bitcoin/pull/31122#discussion_r1814692827), so marking this conversation as resolved.